LabTools/CoreWidgets/UserDataWidget.py

```diff
@@ -0,0 +1,335 @@
+
+        self.bt_rm = QtGui.QPushButton(self)
+        self.bt_rm.setText("Remove last variable")
+        self.bt_rm.clicked.connect(self.bt_rm_click)
+
+        self.addrm_layout.addWidget(self.bt_add)
+        self.addrm_layout.addWidget(self.bt_rm)
+
+
+    def create_notes_textedit(self):
+        self.notes = QtGui.QTextEdit()
+
+
+        self.notes.setPlaceholderText("Notes...")
+
+    def create_footer_layout(self):
+        self.footer_layout = QtGui.QHBoxLayout()
+
+    # Create variable function
+    def new_line(self):
+        text, okPressed = QtGui.QInputDialog.getText(self, "New Data", "Data Name:", QtGui.QLineEdit.Normal, "")
+        if okPressed and text != "":
+            ret = self.create_line(text)
+            if ret is False:
+                QtGui.QMessageBox.information(self, "Unable to create new data", "Data of same name already exists!")
+
+
+    def create_line(self, name):
+        if name not in self.names:
+            widget = SingleLineWidget(self, name, len(self.lines))
+
+            self.lines.append(widget)
+            self.names.append(name)
+            self.line_layout.addWidget(widget)
+
+            return widget
+        else:
+            return False
+
+    def remove_line(self, id):
+        if id >= 0 and id < len(self.lines):
+            self.line_layout.removeWidget(self.lines[id])
+            self.lines[id].close()
+            del self.lines[id]
+            del self.names[id]
+            self.update_lines()
+        else:
+            logging.debug("Incorrect index %s for data line"%str(id))
+    def remove_line_last(self):
+        self.line_layout.removeWidget(self.lines[-1])
+        self.lines[-1].close()
+        del self.lines[-1]
+        del self.names[-1]
+        self.update_lines()
+
+    def remove_line_all(self):
+        for widget in self.lines:
+            self.line_layout.removeWidget(widget)
+            widget.close()
+
+        self.lines = []
+        self.names = ["notes"]
+
+    def update_lines(self):
+        for i, widget in enumerate(self.lines):
+            widget.update_line_number(i)
+    ### EVENTS
+
+    def bt_add_click(self):
+        self.new_line()
+
+    def bt_rm_click(self):
+        self.remove_line_last()
+
+
+    def get_user_data(self):
+        ret = {}
+        for line in self.lines:
+            name, data = line.get_tuple()
+            ret[name] = data
+        if self.notes.toPlainText() != "":
+            ret["notes"] = self.notes.toPlainText()
+        return ret
+
+    def set_user_data(self, obj):
+        self.remove_line_all()
+        for name, data in obj.items():
+            if name == "notes":
+                self.notes.setPlainText(data)
+            else:
+                try:
+                    self.create_line(name).set_data(data)
+                except:
+                    logging.debug("Two data points of same name %s?"%name)
+        self.update_lines()
+
+
+### NOT USED
```

Comment:
if some code is not used, then it can be deleted as we'll always have access to it through the version control

Reply:
I added it here to conform to similar widgets, as it can be used in the same manner, which is why I wrote and left the function there